[`ruff`] improve handling of intermixed comments inside from-imports

amyreese opened the pull request:
Resolves the crash when attempting to format code like:

```
from x import (a as # whatever
b)
```

And chooses to format it as:
```
from x import (
    a as b,  # whatever
)
```

Fixes issue #19138

MichaReiser reviewed (changes requested):
Thanks for looking into this. We'll need to add tests (see `resources/fixtures`). I also suggest playing with different comment combinations. E.g. the following now flips the order of comments which is always undesired:

```py
from x import (a # more 
 as # whatever
b)
```

Gets formatted to

```py
from x import (
    a as b,  # more
    # whatever
)
```

This one doesn't flip the comment order but it places the `other` comment that used to be before `b` after `b`

```py

from x import (a
 as # whatever
 # other
b)

```

Not suggesting that you should or that it makes your life easier in any way. But an alternative fix is to update `place_comment` so that it makes comments between `as` and the first alias leading comments of the first alias

amyreese commented:
In this given example putting comments in every possible position, Black will maintain placement of all the comments, rather than re-flowing any elements together and merging comments:

```
# alpha
from x import (
    # bravo
    a  # charlie
    # delta
    as  # echo
    # foxtrot
    b  # golf
    # hotel
    ,  # india
    # juliet
)  # kilo
```

Do we want to ensure that we follow Black's style and preserve all of these comments in their original positions? Or do we want to have an opinion and re-flow/merge various comments to ensure that `a as b,` is always on one line when possible?

MichaReiser reviewed (commented):
Does this change reflow comments of already formatted code?

Could we do something similar to the `MatchAs` formatting

https://github.com/astral-sh/ruff/blob/63b3f7adae7fcc8910baa902a2d20bfdc4a28ad9/crates/ruff_python_formatter/src/pattern/pattern_match_as.rs#L27-L44

I'm asking because the formatting of 

```py
from x import (a # more 
 as # whatever
b)


from x import (a
 as # whatever
 # other
b)
```

to 

```py
from x import (
    a as b,  # more
    # whatever
)


from x import (
    a as b,  # whatever
    # other
)
```

or

```py
# alpha
from x import (
    # bravo
    a
    
    as  # echo
    # foxtrot
    b  # golf
    # hotel
)
```

where the `foxdrot` comment now comes after `b` (it was a leading comment but it now becomes a trailing comment and the indent for `golf` is off)

```py
# alpha
from x import (
    # bravo
    a as b,  # echo
    # foxtrot
      # golf
    # hotel
)
```


The indent of `# golf` is problematic because this is now a case where the formatter needs two passes to reach convergence. We need to fix this at least and add them as test cases.

amyreese commented:
Documented placement functions and changes, as well as in the alias formatter.  Also added more test cases and updated the logic to better match black style and merge lines/comments when no own-line comments are between the alias and the trailing comma:

Eg,
```py
from foo import (
    bar  # one
    ,  # two
```

Becomes:
```py
from foo import (
    bar,  # one  # two
)
```
